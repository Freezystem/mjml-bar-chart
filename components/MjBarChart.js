import { BodyComponent } from 'mjml-core'

export default class MjBarChart extends BodyComponent {
  static endingTag = true
  
  static dependencies = {
    "mj-column": ["mj-bar-chart"],
    "mj-bar-chart": []
  }

  static allowedAttributes = {
    "title": "string",
    "datasets-label": "string",
    "datasets": "string",
    "groups": "string",
    "colors": "string",
    "axis-color": "color",
    "height": "unit(px)",
    "bar-width": "unit(px)",
    "separator-width": "unit(px)"
  }

  static defaultAttributes = {
    "axis-color":"#d4d4d4",
    "height": "200",
    "bar-width": "30",
    "separator-width": "30"
  }

  #getChartTitle() {
    if (!this.title) {
      return null;
    } 
    
    return `
      <table ${this.htmlAttributes({style: "chartTitleWrapper"})}>
        <tr>
          <td ${this.htmlAttributes({style: "chartTitle"})}>${this.title}</td>
        </tr>
      </table>
    `;
  }

  #getChartBar(value, color) {
    const plainPartHeight = Math.round(value / this.higherValue * this.chartHeight);
    const emptyPartHeight = this.chartHeight - plainPartHeight;
    
    const emptyCellStyle = {
      "padding": "0",
      "height": `${emptyPartHeight}px`
    }

    const plainCellStyle = {
      "padding": "0",
      "height": `${plainPartHeight}px`,
      "background-color": color
    }
    
    return `
      <td style="padding:0">
        <table ${this.htmlAttributes({ style: "chartBarWrapper" })}>
          <tr>
            <td ${this.htmlAttributes({ style: emptyCellStyle })}></td>
          </tr>
          <tr>
            <td ${this.htmlAttributes({ style: plainCellStyle })}></td>
          </tr>
        </table>
      </td>
    `;
  }
  
  #getChartBarSeparator = () => `<td ${this.htmlAttributes({ style: "chartBarSeparator" })}></td>`;
  
  #getChart() {
  	const bars = [this.#getChartBarSeparator()];
    
  	for (const dataset of this.datasets) {
    	for (let i = 0; i < dataset.data.length; i++) {
      	bars.push(this.#getChartBar(dataset.data[i], this.colors[i]));
      }
      
      bars.push(this.#getChartBarSeparator());
    }
    
    return `
      <table ${this.htmlAttributes({ style: "barChart" })}>
        <tr>
          <td style="padding:0;height:10px;"></td>
        </tr>
        <tr>
          ${bars.join("\n")}
        </tr>
      </table>
    `;
  }
  
  #getLabel = value => `<td ${this.htmlAttributes({ style: "chartLabel" })}>${value}</td>`
  
  #getChartLabels() {
    const labels = [this.#getChartBarSeparator()];
    
  	for (const dataset of this.datasets) {
      labels.push(this.#getLabel(dataset.label));
      labels.push(this.#getChartBarSeparator());
    }
    
    return `
      <table ${this.htmlAttributes({ style: "chartLabelWrapper" })}>
        <tr>
          ${labels.join("\n")}
        </tr>
      </table>
    `;
  }
  
  #getLegend(value, color) {
    const legendStyle = {
      "padding": "0 10px",
      "height": "20px",
      "font-size": "14px",
      "border-left": `${this.barWidth}px solid ${color}`
    }
    
    return `
      <span ${this.htmlAttributes({ style: legendStyle })}>${value}</span>
    `;
  }
  
  #getChartLegend() {
    return `
      <table align="center" style="border-collapse:collapse;">
        <tr>
          <td style="padding:0;height:10px;"></td>
        </tr>
        <tr>
          <td style="padding:0;">
            <p ${this.htmlAttributes({ style: "legends" })}>
              ${this.groups.map((g, i) => this.#getLegend(g, this.colors[i]))}
            </p>
          </td>
        </tr>
      </table>
    `;
  }
  
  #getScale() {
  	/* const scale = document.createElement("table");
    scale.style.borderCollapse = "collapse";
  	const top = document.createElement("tr");
    const topOffset = document.createElement("td");
    topOffset.style.padding = "0 5px 0 0";
    topOffset.style.height = "50px";
    topOffset.style.verticalAlign = "bottom";
    topOffset.style.textAlign = "right";
    topOffset.style.color = this.axisColor;
    topOffset.append(`${this.higherValue}`);
    top.append(topOffset);
    
    
  	const middle = document.createElement("tr");
    const middleOffset = document.createElement("td");
    middleOffset.style.padding = "0 5px 0 0";
    middleOffset.style.height = `${this.height + 2}px`;
    middleOffset.style.verticalAlign = "bottom";
    middleOffset.style.textAlign = "right";
    middleOffset.style.color = this.axisColor;
    middleOffset.append("0");
    middle.append(middleOffset);
    
    const bottom = document.createElement("tr");
    const bottomOffset = document.createElement("td");
    bottomOffset.style.padding = "0";
    bottomOffset.style.height = `${20 * this.metrics.length + 60}px`;
    bottom.append(bottomOffset);
    
    scale.append(top, middle, bottom); */
    
    return `
      <table style="border-collapse:collapse;"></table>
    `;
  }

  getStyles() {
    return {
      chartTitleWrapper: {
        "width": "100%",
        "border-collapse": "collapse"
      },
      chartTitle: {
        "padding": "0",
        "height": "40px",
        "font-weight": "bold",
        "text-align": "center",
        "font-size": "20px"
      },
      chartBarSeparator: {
        "padding": "0",
        "min-width": `${this.separatorWidth}px`,
        "max-width": `${this.separatorWidth}px`
      },
      chartBarWrapper: {
        "padding": "0",
        "min-width": `${this.barWidth}px`,
        "max-width": `${this.barWidth}px`
      },
      barChart: {
        "border-collapse": "collapse",
        "border-left": `2px solid ${this.axisColor}`,
        "border-bottom": `2px solid ${this.axisColor}`
      },
      chartLabelWrapper: {
        "border-collapse": "collapse",
        "border-left": "2px solid transparent"
      },
      chartLabel: {
        "height": "30px",
        "padding": "0",
        "font-size": "14px",
        "text-align": "center",
        "overflow": "hidden",
        "min-width": `${this.barWidth * this.groups.length}px`,
        "max-width": `${this.barWidth * this.groups.length}px`
      },
      legends: {
        "margin": "0",
        "padding": "0",
        "max-width": `${this.chartWidth}px`,
        "line-height": "20px",
        "text-align": "center"
      }
    }
  }

  render() {
    this.title = this.getAttribute("title");
    this.datasetValues = this.getAttribute("datasets").split(",").map(v => v.split(" "));
    this.datasets = this.getAttribute("dataset-labels")
    .split(",")
    .map((label, idx) => ({
      label,
      data: this.datasetValues[idx].map(n => parseInt(n, 10))
    }));
    this.higherValue = this.datasetValues.reduce((vs, d) => [...vs, ...d], []).sort((a, b) => b - a)[0];
    this.colors = this.getAttribute("colors").split(",");
    this.axisColor = this.getAttribute("axis-color");
  	this.groups = this.getAttribute("groups").split(",");
    this.chartHeight = parseInt(this.getAttribute("height"), 10);
    this.barWidth = parseInt(this.getAttribute("bar-width"), 10);
    this.separatorWidth = parseInt(this.getAttribute("separator-width"), 10);
    this.chartWidth = 2 + (this.separatorWidth * (this.groups.length + 1)) +
      (this.barWidth * this.datasets.length * this.groups.length);

    return `
      <table style="border-collapse:collapse;">
        <tr>
          <td style="padding:0;">
            ${this.#getScale()}
            <table style="border-collapse:collapse;">
              <tr>
                <td style="padding:0;">
                  ${this.#getChartTitle()}
                </td>
              </tr>
              <tr>
                <td style="padding:0;">
                  ${this.#getChart()}
                </td>
              </tr>
              <tr>
                <td style="padding:0;">
                  ${this.#getChartLabels()}
                </td>
              </tr>
              <tr>
                <td style="padding:0;">
                  ${this.#getChartLegend()}
                </td>
              </tr>
            </table>
          </td>
        </tr>
      </table>
    `;
  }
}
