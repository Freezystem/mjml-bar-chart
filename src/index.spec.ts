import mjml2html from "mjml";
import { registerComponent } from "mjml-core";
import MjBarChart from "./index";

function toHtml(mjml: string): string {
	const { html, errors } = mjml2html(mjml);
	return errors.length > 0 ? errors[0].message : html;
}

describe("mjml-bar-chart", () => {
	beforeAll(() => {
		registerComponent(MjBarChart);
	});

	describe("mjml markup", () => {
		it("should render the bar chart", () => {
			const mjml = `
			  <mjml>
				<mj-body>
				  <mj-section>
					<mj-column>
					  <mj-bar-chart
						title="Sum of Requests by Department"
						dataset-labels="January,February,March" 
						datasets="[[33,14,27],[18,66,42],[7,15,21]]"
						groups="support,sales,tech"
						colors="#ffe5ec,#ffb3c6,#fb6f92"/>
					</mj-column>
				  </mj-section>
				</mj-body>
			  </mjml>
			`;

			expect(toHtml(mjml)).toMatchSnapshot();
		});
	});

	describe("getChartBarSeparator", () => {
		it("should render chart bar separator", () => {
			const barChart = new MjBarChart({
				attributes: {
					"dataset-labels": "January,February,March",
					datasets: "[[33,14,27],[18,66,42],[7,15,21]]",
					groups: "support,sales,tech",
					colors: "#ffe5ec,#ffb3c6,#fb6f92",
				},
			});

			const gen = barChart["getChartBarSeparator"]();

			expect(gen).toBe('<td style="padding:0;min-width:30px;max-width:30px;"></td>');
		});
	});
});
