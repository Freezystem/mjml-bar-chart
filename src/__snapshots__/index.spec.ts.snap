// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mjml-bar-chart mjml markup should render the bar chart 1`] = `
"<!doctype html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office">
  <head>
    <title></title>
    <!--[if !mso]><!-->
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <!--<![endif]-->
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type="text/css">
      #outlook a { padding:0; }
      body { margin:0;padding:0;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%; }
      table, td { border-collapse:collapse;mso-table-lspace:0pt;mso-table-rspace:0pt; }
      img { border:0;height:auto;line-height:100%; outline:none;text-decoration:none;-ms-interpolation-mode:bicubic; }
      p { display:block;margin:13px 0; }
    </style>
    <!--[if mso]>
    <noscript>
    <xml>
    <o:OfficeDocumentSettings>
      <o:AllowPNG/>
      <o:PixelsPerInch>96</o:PixelsPerInch>
    </o:OfficeDocumentSettings>
    </xml>
    </noscript>
    <![endif]-->
    <!--[if lte mso 11]>
    <style type="text/css">
      .mj-outlook-group-fix { width:100% !important; }
    </style>
    <![endif]-->
    
    
    <style type="text/css">
      @media only screen and (min-width:480px) {
        .mj-column-per-100 { width:100% !important; max-width: 100%; }
      }
    </style>
    <style media="screen and (min-width:480px)">
      .moz-text-html .mj-column-per-100 { width:100% !important; max-width: 100%; }
    </style>
    
  
    <style type="text/css">
    
    
    </style>
    <style type="text/css">
    
    </style>
    
  </head>
  <body style="word-spacing:normal;">
    
    
      <div
         style=""
      >
        
      
      <!--[if mso | IE]><table align="center" border="0" cellpadding="0" cellspacing="0" class="" role="presentation" style="width:600px;" width="600" ><tr><td style="line-height:0px;font-size:0px;mso-line-height-rule:exactly;"><![endif]-->
    
      
      <div  style="margin:0px auto;max-width:600px;">
        
        <table
           align="center" border="0" cellpadding="0" cellspacing="0" role="presentation" style="width:100%;"
        >
          <tbody>
            <tr>
              <td
                 style="direction:ltr;font-size:0px;padding:20px 0;text-align:center;"
              >
                <!--[if mso | IE]><table role="presentation" border="0" cellpadding="0" cellspacing="0"><tr><td class="" style="vertical-align:top;width:600px;" ><![endif]-->
            
      <div
         class="mj-column-per-100 mj-outlook-group-fix" style="font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;"
      >
        
      <table
         border="0" cellpadding="0" cellspacing="0" role="presentation" style="vertical-align:top;" width="100%"
      >
        <tbody>
          
              <tr>
                <td
                   style="font-size:0px;word-break:break-word;"
                >
                  
          <table class="mjmlBarChart" style="border-collapse:collapse;margin:0 auto;">
            <tr>
              
          <td style="padding:0;vertical-align:top;">
            <table style="border-collapse:collapse;">
              
                <tr>
                  <td style="padding:0 5px 0 0;height:56px;vertical-align:bottom;text-align:right;font-size:14px;color:d4d4d4;">66</td>
                </tr>
              

                <tr>
                  <td style="padding:0 5px 0 0;height:50.5px;vertical-align:bottom;text-align:right;font-size:14px;color:d4d4d4;">49</td>
                </tr>
              

                <tr>
                  <td style="padding:0 5px 0 0;height:50.5px;vertical-align:bottom;text-align:right;font-size:14px;color:d4d4d4;">33</td>
                </tr>
              

                <tr>
                  <td style="padding:0 5px 0 0;height:50.5px;vertical-align:bottom;text-align:right;font-size:14px;color:d4d4d4;">16</td>
                </tr>
              

                <tr>
                  <td style="padding:0 5px 0 0;height:50.5px;vertical-align:bottom;text-align:right;font-size:14px;color:d4d4d4;">0</td>
                </tr>
              
            </table>
          </td>
        
              
          <td style="padding:0;">
            <table style="border-collapse:collapse;">
              
          <tr>
            <td style="padding:0;">
              <table style="width:100%;border-collapse:collapse;">
                <tr>
                  <td style="padding:0;height:40px;font-weight:bold;text-align:center;font-size:20px;">Sum of Requests by Department</td>
                </tr>
              </table>
            </td>
          </tr>
        
              
          <tr>
            <td style="padding:0;">
              <table style="border-collapse:collapse;border-left:2px solid d4d4d4;border-bottom:2px solid d4d4d4;">
                <tr>
                  <td style="padding:0;min-width:30px;max-width:30px;"></td>

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:116px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">33</td>
              </tr>
              <tr>
                <td style="padding:0;height:100px;background-color:#ffe5ec;"></td>
              </tr>
            </table>
          </td>
        

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:174px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">14</td>
              </tr>
              <tr>
                <td style="padding:0;height:42px;background-color:#ffb3c6;"></td>
              </tr>
            </table>
          </td>
        

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:134px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">27</td>
              </tr>
              <tr>
                <td style="padding:0;height:82px;background-color:#fb6f92;"></td>
              </tr>
            </table>
          </td>
        
<td style="padding:0;min-width:30px;max-width:30px;"></td>

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:161px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">18</td>
              </tr>
              <tr>
                <td style="padding:0;height:55px;background-color:#ffe5ec;"></td>
              </tr>
            </table>
          </td>
        

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:16px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">66</td>
              </tr>
              <tr>
                <td style="padding:0;height:200px;background-color:#ffb3c6;"></td>
              </tr>
            </table>
          </td>
        

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:89px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">42</td>
              </tr>
              <tr>
                <td style="padding:0;height:127px;background-color:#fb6f92;"></td>
              </tr>
            </table>
          </td>
        
<td style="padding:0;min-width:30px;max-width:30px;"></td>

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:195px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">7</td>
              </tr>
              <tr>
                <td style="padding:0;height:21px;background-color:#ffe5ec;"></td>
              </tr>
            </table>
          </td>
        

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:171px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">15</td>
              </tr>
              <tr>
                <td style="padding:0;height:45px;background-color:#ffb3c6;"></td>
              </tr>
            </table>
          </td>
        

          <td style="padding:0">
            <table style="padding:0;min-width:30px;max-width:30px;">
              <tr>
                <td style="padding:0;height:152px;font-size:12px;vertical-align:bottom;text-align:center;line-height:16px;">21</td>
              </tr>
              <tr>
                <td style="padding:0;height:64px;background-color:#fb6f92;"></td>
              </tr>
            </table>
          </td>
        
<td style="padding:0;min-width:30px;max-width:30px;"></td>
                </tr>
              </table>
            </td>
          </tr>
        
              
          <tr>
            <td style="padding:0;">
              <table style="border-collapse:collapse;border-left:2px solid transparent;">
                <tr>
                  <td style="padding:0;min-width:30px;max-width:30px;"></td>
<td style="height:30px;padding:0;font-size:14px;text-align:center;overflow:hidden;min-width:90px;max-width:90px;">January</td>
<td style="padding:0;min-width:30px;max-width:30px;"></td>
<td style="height:30px;padding:0;font-size:14px;text-align:center;overflow:hidden;min-width:90px;max-width:90px;">February</td>
<td style="padding:0;min-width:30px;max-width:30px;"></td>
<td style="height:30px;padding:0;font-size:14px;text-align:center;overflow:hidden;min-width:90px;max-width:90px;">March</td>
<td style="padding:0;min-width:30px;max-width:30px;"></td>
                </tr>
              </table>
            </td>
          </tr>
        
              
          <tr>
            <td style="padding:0;">
              <table style="border-collapse:collapse;width:100%;">
                <tr>
                  <td style="padding:0;height:10px;"></td>
                </tr>
                <tr>
                  <td style="padding:0;">
                    <p style="margin:0;padding:0;max-width:392px;line-height:20px;text-align:center;">
                      <span style="padding:0 10px;height:20px;font-size:14px;border-left:30px solid #ffe5ec;">support</span>,<span style="padding:0 10px;height:20px;font-size:14px;border-left:30px solid #ffb3c6;">sales</span>,<span style="padding:0 10px;height:20px;font-size:14px;border-left:30px solid #fb6f92;">tech</span>
                    </p>
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        
            </table>
          </td>
        
            </tr>
          </table>
        
                </td>
              </tr>
            
        </tbody>
      </table>
    
      </div>
    
          <!--[if mso | IE]></td></tr></table><![endif]-->
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    
      
      <!--[if mso | IE]></td></tr></table><![endif]-->
    
    
      </div>
    
  </body>
</html>
  "
`;
